test_printf/test_00.cl

```diff
@@ -0,0 +1,2 @@
+// Options: -cl-std=CL3.0
```

Comment:
Looks like the names of the `test_*.cpp` sources in https://github.com/KhronosGroup/OpenCL-CTS are not unique. So I see seven sources named `test_loop.cpp` in various parts of the repo. Would it be a good idea to name this to be something like `printf/test_printf/test_00.cl`?

Also a bit of concern regarding the naming of the kernel sources themselves. Apart from the question on whether it's known that two digits would be enough for all series of tests we might have here, what is supposed to happen if we remove some test cases or add new ones that would get numbers somewhere in the middle of the series? Ideally, maybe it could be something like `test-5d.cl`, though clearly this suggests some more work updating the test generators to give tests cases meaningful, unique and persistent ids.

Reply:
> Would it be a good idea to name this to be something like printf/test_printf/test_00.cl

You mean preserve existing CTS directory layout as well? This idea LGTM.

> Apart from the question on whether it's known that two digits would be enough for all series of tests we might have here, what is supposed to happen if we remove some test cases or add new ones that would get numbers somewhere in the middle of the series?

Yeah, I am not sure yet how to address this properly. Originally, I was thinking that simply appending the name of a new test with a new digit won't hurt: we don't really have to care about the order of uploaded kernels. But I didn't think of removing existing tests... As an alternative, it may be reasonable to provide a hash within a test name. I used opencl-intercept-layer (https://github.com/intel/opencl-intercept-layer) and originally there was a hash within a name. This is a reliable solution, but it may cause difficulties is support, if, for example, other kernel dumpers use different policy.

Reply:
I remember that using the dumping SHA was discussed but there were some concerns whether they are stable enough.

Do you know by chance what info is used to generate those?

CC @StuartDBrady

Reply:
> Do you know by chance what info is used to generate those?

I am not sure, but I think it somehow uses kernel source code...

> I remember that using the dumping SHA was discussed but there were some concerns whether they are stable enough.

To make it stable, maybe we could introduce a script and include it in this repository. I imagine that would be very straightforward: pick kernel source code, strip out all white spaces, convert to lower/uppercase and then use SHA1 or MD5.

Reply:
Yes: almost every test in the CTS will dump CL source when the `--compilation-mode` option is used.  See the common `--help` text.

How was this dump produced?  If it was produced by manually copying and pasting kernel strings from the CTS's C++ source, then that does not seem sustainable to me.

I had proposed that we would use files dump with `--compilation-mode`, but with the second hash in the filename removed for each CL source file.  (The second hash is the hash of the compilation options, but this has no place in the filename of the source unless the options are included in the source file - I believe the options are excluded from the CL source file when dumped by the CTS, and are instead written to a separate ".options" file.)

Reply:
FWIW, the "hash" function used by the CTS is actually a CRC32 (specifically, the Ethernet FCS CRC) which is a very poor choice for a hash function (owing to its lack of diffusion and avalanche effect).

MD5 would be a better choice:
* there are numerous public domain implementations available
* it's generally very well-supported in OSes, etc.
* implementations tend to be fast

(There are hashes for which faster software implementations are possible, e.g. SHA1, but there are not as widespread as MD5. Note that neither MD5 nor SHA1 should be used where cryptographic strength is required, but that is not what is being discussed here.)

Reply:
Yes, a proper hash would be an improvement. Might be even better with some meaningful names, like this:

https://github.com/kosarev/OpenCL-CTS/commit/7005d51802a01e1d17698043bf2a37d7688840da

A few side notes:
* It seems the formatting of the sources in this pull request differs from what we originally get with `--compilation-cache-mode`, which might complicate updating this repository as well as making sure it's correct and complete. I think it is so for more or less any manual effort we may apply to the sources. So if we are after some better-looking formatting, then maybe we could just fix the generators in the CTS repo?
* Similarly, I wonder if it would make sense to ask the CTS to direct the options to `// OPTIONS:` lines in the resulting sources instead of creating another bunch of `*.options` files?
* Also, is there any particular reason why with the `--compilation-cache-mode` option in place we still actually run the tests on the target?
* It might be not the case for other groups of tests, but for the `printf` ones the resulting sources are times less than its generator. Wouldn't it be easier to just have the sources as... well, the actual source of the kernels code -- along with the expected results?

Reply:
> How was this dump produced? If it was produced by manually copying and pasting kernel strings from the CTS's C++ source, then that does not seem sustainable to me.

I used opencl-intercept-layer (https://github.com/intel/opencl-intercept-layer) and managed to get the same kernel sources as with `--compilation-cache-mode dump-cl-files`. 

> It seems the formatting of the sources in this pull request differs from what we originally get with --compilation-cache-mode, which might complicate updating this repository as well as making sure it's correct and complete. I think it is so for more or less any manual effort we may apply to the sources. So if we are after some better-looking formatting, then maybe we could just fix the generators in the CTS rep

> Similarly, I wonder if it would make sense to ask the CTS to direct the options to // OPTIONS: lines in the resulting sources instead of creating another bunch of *.options files

I am not much familiar with CTS dumping policy. If it will be decided to fix formatting and options dumping right in `OpenCL-CTS` repo that's perfectly fine, I think. But I don't think that should affect this repository in all cases, except if we would like to reuse kernels from here right in CTS (see below).

> It might be not the case for other groups of tests, but for the printf ones the resulting sources are times less than its generator. Wouldn't it be easier to just have the sources as... well, the actual source of the kernels code -- along with the expected results

I think by this you meant that we could reuse kernels from `OpenCL-CTS-Kernel-Dumps` right in `OpenCL-CTS`, right? I think that is what we should strive to achieve: when ` OpenCL-CTS-Kernel-Dumps` gets more mature, we can later get rid of all the kernel strings from C++ files in `OpenCL-CTS` and preopload kernels from `OpenCL-CTS-Kernel-Dumps` instead.  We should come up with such format to make it reusable for this purpose.

Reply:
> I used opencl-intercept-layer (https://github.com/intel/opencl-intercept-layer) and managed to get the same kernel sources as with `--compilation-cache-mode dump-cl-files`.

One feature of `--compilation-cache-mode dump-cl-files` seems to be that it uses the actual kernel names to name the generated source files, which, together with preserving the original formatting, may help keep things reproducible, meaning making life a bit easier for those who will update the dumped sources to reflect changes in the CTS repo, reviewers including. That is also why the patch for the `printf()` tests mentioned above, when used with the command-line option, looks like it may improve the situation.

I guess that custom intercepting machinery could do the same that option does, but is it worth the effort preparing and publishing the intercepting code given everything we need is already there in the CTS code base?

> I think that is what we should strive to achieve: when ` OpenCL-CTS-Kernel-Dumps` gets more mature, we can later get rid of all the kernel strings from C++ files in `OpenCL-CTS` and preopload kernels from `OpenCL-CTS-Kernel-Dumps` instead.

Perfect. So sounds like we do want this repo to match the latest state of the CTS code -- means regular updates.

So I think my ultimate question is, do we see the work on this repo as lots of manual effort generating/formatting/committing/reviewing bunches of sources or rather as building an infrastructure for updating it, probably completely automatically or at least having it all reproducible locally, maybe even without access to/use of any GPU hardware, which technically seems to be perfectly doable, doesn't it?

Reply:
With all that being discussed, I suggest fixing the PR with following:

- Preserve directories layout as in `OpenCL-CTS`
- Use `--compilation-cache-mode dump-cl-files` to dump kernels along with the options. 
- Submit kernel files together with option files with a tiny fix to make kernel source contain options string with comment `// Options:`

Another concern which I have here is that for some CTS tests the amount of files produces if really huge, for example as for `test_c11_atomics`. Pretty all the kernels follow the same pattern, but still, should they be submitted here as is?

Reply:
Do we know how many kernels they are? I would say the original intent was to gether all kernels but if we reach scalability issues then we can reconsider the alternatives.

Reply:
> So I think my ultimate question is, do we see the work on this repo as lots of manual effort generating/formatting/committing/reviewing bunches of sources or rather as building an infrastructure for updating it, probably completely automatically or at least having it all reproducible locally, maybe even without access to/use of any GPU hardware, which technically seems to be perfectly doable, doesn't it?

@kosarev I think this might be a good longer term goal. However I imagine once we get it working we should probably integrate it into the OpenCL-CTS repo directly. But we can build and experiment with it here.

For now, the minimal requirement is to gather as many kernels as we can so that we can help compiler tools to improve the testing offline.

Reply:
> Do we know how many kernels they are? I would say the original intent was to gether all kernels but if we reach scalability issues then we can reconsider the alternatives.

If I did everything correct, the dump directory contains 28579 kernel files. The overall size is somewhat like 278 MB...

Reply:
@AnastasiaStulova Sure, just not giving ourselves extra work with sequential numbering, manual formatting, etc. should be fine for now.

> With all that being discussed, I suggest fixing the PR with following:
>     * Preserve directories layout as in `OpenCL-CTS`
>     * Use `--compilation-cache-mode dump-cl-files` to dump kernels along with the options.
>     * Submit kernel files together with option files with a tiny fix to make kernel source contain options string with comment `// Options:`

@azabazno Sounds good to me. Would be nice to have that tiny fix published somewhere at some point, maybe preferably have it changed in the CTS repo, if we can.

Reply:
> If I did everything correct, the dump directory contains 28579 kernel files. The overall size is somewhat like 278 MB...

@azabazno What suite did you ran?
With `python run_conformance.py opencl_conformance_tests_quick.csv`, I got 46408 kernel files.
I use the `--compilation-cache-mode dump-cl-files` to dump the kernels.

Reply:
> > If I did everything correct, the dump directory contains 28579 kernel files. The overall size is somewhat like 278 MB...
> 
> @azabazno Which suite did you use?
> With `python run_conformance.py opencl_conformance_tests_quick.csv`, I got 46408 kernel files.
> I use the `--compilation-cache-mode dump-cl-files` to dump the kernels.

I got 28579 kernels for `test_c11_atomics` only.